### DISCUSSION TITLE
Polygon-SDK accepts transactions with an invalid gas limit as valid transactions

### DISCUSSION THREAD

**Discussion by AzulPretoBranco** [0xPolygon/polygon-edge#294] [Ideas] [answered]
### Description

During our internal testing we notice that if we submit a transaction with an invalid gas limit (this was done on purpose) to a blockchain created using Polygon-SDK the transaction is accepted by the blockchain and added to the tx pool. 

Later when the method `writeTransactions`runs the blockchain will check the gas limit and if the value is invalid the transaction will be silently discarded.

This poses a problem, because the transaction is silently discarded applications that use the blockchain (like metamask) will not be able to receive a confirmation for this transaction meaning that the transaction will be "forever pending".

This exact use case doesn't happen in an ethereum test network like the Rinkeby Test Network. The ethereum blockchain validates the gas limit before adding the transaction to the tx pool and returns an appropriate error to the application interacting with the blockchain (like metamask).

Attached is a video that shows this test using the Rinkeby Test Network and a blockchain created using the Polygon-SDK.

### Your environment

* OS and version: Windows 10 (WSL Ubuntu 20.04.3 LTS
* version of the Polygon SDK: develop branch at commit 7fe8091967102192539618fd786f226d54dd312b

### Steps to reproduce

Video that shows the use cases in the issue description https://1drv.ms/u/s!AvhuC05QeOJntn9-aiUsXaKgKqZN

### Expected behaviour

We believe the Polygon-SDK should behave similar to the ethereum blockchain and return an error when we submit an invalid transaction.

### Actual behaviour

The Polygon-SDK accepts a transaction with an invalid gas limit as a valid transaction unlike the ethereum network.

### Logs

Logs from Polygon-SDK:
```
2021-12-15T16:23:52.952Z [INFO]  polygon.consensus.ibft.acceptState: we are the proposer: block=71149
2021-12-15T16:23:52.952Z [ERROR] polygon.consensus.ibft: failed to write transaction: transaction's gas limit exceeds block gas limit
2021-12-15T16:23:52.952Z [INFO]  polygon.consensus.ibft: picked out txns from pool: num=0 remaining=0
```

### Proposed solution

Maybe do something similar to go-ethereum where they have the check in `validateTx`, screenshot from go-ethereum:
![image](https://user-images.githubusercontent.com/96196879/146232789-850e26c6-ecfa-40c2-a509-148e2616b745.png)

**Comment by zivkovicmilos** [comment, marked as answer]
Hey @AzulPretoBranco,

Thank you for reaching out.

So far what you've described is normal behavior - the transaction is accepted into the TxPool of the Polygon SDK but is discarded when it is attempted to be applied. 

This is because block limits are enforced during transaction execution when the block is being built, and not at the point where you're adding the transaction - which may or may not succeed for whatever reason.

The reason block limits are not checked in transaction addition is because these are consensus specific - the TxPool shouldn't care about whether or not the transaction can be executed if it passes the basic checks to be in the pool in the first place.

Apps like Metamask, or any web3 providers in general query the transaction by hash or attempt to get the receipt of the transaction after submitting it, and update their own status according to the responses.

I'll sync with the rest of the team so we can find an optimal solution to this pending transaction status in MM.

I'll keep you posted on this, moving this to a discussion since it's not necessarily an issue with the Polygon SDK, but rather a feature request 👍

**Reply by AzulPretoBranco** [reply]
Hi @zivkovicmilos,

Thanks for the explanation!
I understand your point, but in that case I have a question that maybe you can answer:

What is the optimal way (if there is any) for us to detect transactions that where discarded by the chain for error reasons similar to this one?

As you said, typically what apps do is check the transaction receipt to see if the status was success or failure, but in this case, from what I understand, the Polygon SDK blockchain will never generate a transaction receipt.

We can query the transaction by the hash but, correct me if I'm wrong, that will not return anything we can use to check if the transaction was discarded or not.

Do you have any suggestions?

**Reply by 0xAleksaOpacic** [reply]
Hi @AzulPretoBranco 
We have finally merged the solution for this problem in develop branch and it will be released in the next release.

Here we had 2 problems as you have described in the issue.
1. We didn't check if the transaction is valid (tx.gasLimit exceeds BlockGasLimit) before putting it in the pool
2. When writing transactions, transactions will be discarded silently if tx.gasLimit exceeds BlockGasLimit

**Why we had a problem in a first-place**
We do not have a default value for block size. Every time we build the block, we calculate the gas limit dynamically. That being said when a new transaction arrives in the pool we do not what will block limit be. Because that block is not yet created. We had to validate does tx.gasLimit exceeds BlockGasLimit when we are building the block. And at that moment clients like metamask expect that we would generate a receipt for that transaction and they stuck waiting for it.

**Solutions**
Even tho we calculate the gas limit dynamically, the value between a few consecutive blocks will be relatively the same. When the transaction arrives we can check **what was the gas limit of the latest block** and expect the value of the block where the transaction will be sealed to be relatively the same. This way we can discard invalid transactions before they enter the pool (same as go-eth) 

Still remains a rare edge case. When the transaction arrives it will be valid for the latest block. But invalid for the block in which it will be sealed. For this we have implemented **failed transaction receipts** which are going to be sealed in the block. So clients will get response with receipt with status false.

**Reply by AzulPretoBranco** [reply]
Hi @Aleksao998 

Thanks so much for the fix and the detailed explanation!
Best regards!

**Comment by jooddang** [comment]
what token do you use for paying gas on "polygon edge" blockchain you created?
I can't find what's the native token of a "polygon edge" blockchain. any help?
